### PR TITLE
deal with null/none connections

### LINKED ISSUE #23621: meta: reset_connection not working
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

`meta` module's `reset_connection`

##### ANSIBLE VERSION

```
[vagrant@localhost infra-bootstrap]$ ansible --version
ansible 2.3.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.5 (default, Sep  5 2016, 02:30:38) [GCC 4.8.5 20150623 (Red Hat 4.8.5-9)]
```

##### OS / ENVIRONMENT

OEL7

##### SUMMARY

Using the `meta: reset_connection` instruction appears to be broken under very simple use case.

##### STEPS TO REPRODUCE

Using a very simple playbook (below), the `meta: reset_connection` instruction stack traces (command below under `ACUAL RESULTS`)

simple `test.yml`:
```yaml
---
- hosts: all
  tasks:
  - meta: reset_connection
```

inventory:
```
# Generated by Vagrant

node01
core ansible_connection=local

[core]
core

[development]
core
```

##### EXPECTED RESULTS

The ssh connection to get reset successfully

##### ACTUAL RESULTS

Ansible stack traces

```
[vagrant@localhost infra-bootstrap]$ ansible-playbook --limit=core --inventory-file=/tmp/vagrant-ansible/inventory -vvv test.yml
Using /etc/ansible/ansible.cfg as config file
 [WARNING]: Found both group and host with same name: core


PLAYBOOK: test.yml ***************************************************************************************************************************************************
1 plays in test.yml

PLAY [all] ***********************************************************************************************************************************************************

TASK [Gathering Facts] ***********************************************************************************************************************************************
Using module file /usr/lib/python2.7/site-packages/ansible/modules/system/setup.py
<core> ESTABLISH LOCAL CONNECTION FOR USER: vagrant
<core> EXEC /bin/sh -c 'echo ~ && sleep 0'
<core> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/vagrant/.ansible/tmp/ansible-tmp-1492207079.55-247174731112807 `" && echo ansible-tmp-1492207079.55-247174731112807="` echo /home/vagrant/.ansible/tmp/ansible-tmp-1492207079.55-247174731112807 `" ) && sleep 0'
<core> PUT /tmp/tmpbuhXwS TO /home/vagrant/.ansible/tmp/ansible-tmp-1492207079.55-247174731112807/setup.py
<core> EXEC /bin/sh -c 'chmod u+x /home/vagrant/.ansible/tmp/ansible-tmp-1492207079.55-247174731112807/ /home/vagrant/.ansible/tmp/ansible-tmp-1492207079.55-247174731112807/setup.py && sleep 0'
<core> EXEC /bin/sh -c '/usr/bin/python /home/vagrant/.ansible/tmp/ansible-tmp-1492207079.55-247174731112807/setup.py; rm -rf "/home/vagrant/.ansible/tmp/ansible-tmp-1492207079.55-247174731112807/" > /dev/null 2>&1 && sleep 0'
ok: [core]
META: ran handlers
ERROR! Unexpected Exception: 'NoneType' object has no attribute 'reset'
the full traceback was:

Traceback (most recent call last):
  File "/usr/bin/ansible-playbook", line 109, in <module>
    exit_code = cli.run()
  File "/usr/lib/python2.7/site-packages/ansible/cli/playbook.py", line 154, in run
    results = pbex.run()
  File "/usr/lib/python2.7/site-packages/ansible/executor/playbook_executor.py", line 153, in run
    result = self._tqm.run(play=play)
  File "/usr/lib/python2.7/site-packages/ansible/executor/task_queue_manager.py", line 284, in run
    play_return = strategy.run(iterator, play_context)
  File "/usr/lib/python2.7/site-packages/ansible/plugins/strategy/linear.py", line 221, in run
    results.extend(self._execute_meta(task, play_context, iterator, host))
  File "/usr/lib/python2.7/site-packages/ansible/plugins/strategy/__init__.py", line 908, in _execute_meta
    connection.reset()
AttributeError: 'NoneType' object has no attribute 'reset'
```

### PR DESCRIPTION
**Pull request by bcoca**
##### SUMMARY

fixes #23621

##### ISSUE TYPE

 - Bugfix Pull Request

##### COMPONENT NAME

strategy

##### ANSIBLE VERSION

```
2.3/2.4
```

##### ADDITIONAL INFORMATION

'smart' wasn't resolved until the task_executor stage, with meta executing at the strategy plugin level it did not get the data it needed at this point. So moving the 'resolution' of smart to the play_context now allows for the resolution to be the same.

### CODE REVIEW

File: lib/ansible/playbook/play_context.py [hunk 1]
```diff
@@ -616,3 +618,29 @@ def update_vars(self, variables):
                         variables[var_opt] = var_val
             except AttributeError:
                 continue
+
+    @property
+    def connection(self):
```
Comment (pilou-) [inline lib/ansible/playbook/play_context.py]: Should `connection` property not be named differently than the `connection` attribute ? It seems the property will be overridden by the `_connection` `FieldAttribute` defined in `Base` and never used/called.

**Review by pilou- — COMMENTED** [review]
The updated code could be kept in `lib/ansible/executor/task_executor.py`.

Line 719, when the condition is True, `conn_type` value is set to `persistent`, should `paramiko` not be used instead ?

**Comment by bcoca** [comment]
@snowsky the connection can be set to 'smart' which does not match a plugin, got resolved in task_executor previouslly, i moved it to the play_context as it makes more sense there.

**Comment by snowsky** [comment]
@bcoca ic, I think this is a better way to fix. +1

**Pull request merged**